src/orange/BoundingBox.hh

```diff
@@ -7,65 +7,81 @@
 //---------------------------------------------------------------------------//
 #pragma once
 
-#include <cmath>
+#include <type_traits>
 
 #include "corecel/Assert.hh"
 #include "corecel/Macros.hh"
 #include "corecel/cont/Array.hh"
 #include "corecel/math/NumericLimits.hh"
 
-#include "Types.hh"
+#include "OrangeTypes.hh"
 
 namespace celeritas
 {
 //---------------------------------------------------------------------------//
 /*!
  * Axis-aligned bounding box.
  *
- * The bounding box can be constructed in an unassigned state, in which lower
- * and upper cannot be called.
+ * The default bounding box is "degenerate", which encloses no space and
+ * evaluates to "false". A degenerate bounding box still has the ability to be
+ * unioned and intersected with other bounding boxes with the expected effect.
  */
 template<class T>
 class BoundingBox
 {
   public:
     //!@{
     //! \name Type aliases
-    using value_type = T;
-    using Real3 = Array<value_type, 3>;
+    using real_type = T;
+    using Real3 = Array<real_type, 3>;
     //!@}
 
+  public:
     // Construct from infinite extents
     static inline CELER_FUNCTION BoundingBox from_infinite();
 
+    // Construct from unchecked lower/upper bounds
+    static CELER_CONSTEXPR_FUNCTION BoundingBox
+    from_unchecked(Real3 const& lower, Real3 const& upper);
+
     // Construct in unassigned state
-    inline CELER_FUNCTION BoundingBox();
+    CELER_CONSTEXPR_FUNCTION BoundingBox();
 
     // Construct from upper and lower points
     inline CELER_FUNCTION BoundingBox(Real3 const& lower, Real3 const& upper);
 
     //// ACCESSORS ////
 
-    // Lower bbox coordinate
-    CELER_FORCEINLINE_FUNCTION Real3 const& lower() const;
+    //! Lower bbox coordinate
+    CELER_CONSTEXPR_FUNCTION Real3 const& lower() const { return lower_; }
+
+    //! Upper bbox coordinate
+    CELER_CONSTEXPR_FUNCTION Real3 const& upper() const { return upper_; }
+
+    // Whether the bbox is nondegenerate
+    CELER_CONSTEXPR_FUNCTION explicit operator bool() const;
 
-    // Upper bbox coordinate
-    CELER_FORCEINLINE_FUNCTION Real3 const& upper() const;
+    //// MUTATORS ////
 
-    // Whether the bbox is assigned
-    CELER_FORCEINLINE_FUNCTION explicit operator bool() const;
+    // Intersect in place with a half-space
+    CELER_CONSTEXPR_FUNCTION void
+    clip(Sense sense, Axis axis, real_type position);
 
   private:
     Real3 lower_;
     Real3 upper_;
+
+    // Implementation of 'from_unchecked' (true type 'tag')
```

Comment:
true type 'tag'?

Reply:
A "tag" struct is a type that's only used to assist in type-dependent overloading, or altering the behavior of a function based on some condition. Here, `std::true_type` is a "tag" that means "do not check".